rplugin/python3/deoplete/sources/deoplete_jedi.py

```diff
@@ -47,21 +57,120 @@ def get_complete_position(self, context):
+
+        Only search for def or class for context.  Jedi only returns results
+        from the lines before the current one.
+        """
+        start = line - 1
+        indent = len(source[start]) - len(source[start].lstrip())
+
+        for i in range(start, 0, -1):
+            s_line = source[i]
+            if not s_line.strip() or not _block_re.match(s_line):
+                continue
+
+            line_indent = len(s_line) - len(s_line.lstrip())
+            if line_indent < indent:
+                start = i
+                indent = line_indent
+                break
+
+        return [start, line - 1]
+
     def gather_candidates(self, context):
         line = context['position'][1]
         col = context['complete_position']
         buf = self.vim.current.buffer
-        source = '\n'.join(buf[:])
+        source = buf[:]
         cline = self.vim.current.line
 
+        extra_modules = []
+        cache_key = None
+        deoplete_input = context['input'].strip()
+        cache_line = 0
+
+        if self.cache_enabled:
+            # Caching based on context input.  If the input is blank, it was
+            # triggered with `.` to get module completions.
+            #
+            # The module files as reported by Jedi are stored with their
+            # modification times to help detect if a cache needs to be
+            # refreshed.
+            #
+            # For scoped variables in the buffer, construct a cache key using
+            # the filename and a partial input prefix.  The buffer file's
+            # modification time is checked to see if the completion needs to be
+            # refreshed.  The approximate scope lines are also cached to help
+            # invalidate the cache based on line position.
+
+            if context.get('complete_str'):
+                # Note: Module completions will be an empty string.
+                #
+                # Jedi filters the results making making the cache impractical
+                # for buffer variables.  Force Jedi to complete a partial
+                # string based on deoplete.nvim's minimum start length.
+                #
+                # Typing `ose` will return only return matches that start with
+                # `ose.*`.  Typing `os` will return matches that start with
+                # `os.*`, which includes `ose.*`.  Without forcing the minimum
+                # length, we would end up too many cache entries with
+                # duplicates.
+                #
+                # Another reason is deoplete.nvim does fuzzy matching.  Without
+                # this, completing `oseo` would give no results from Jedi.
+                # With the partially cached `os` entry, it will be able to
+                # match `OSError`.
+                word = context.get('complete_str')
+                cache_key = '{}.{}'.format(buf.name,
+                                           word[:self.complete_min_length].lower())
```

Comment:
I always ignore E501 unless it's a very long line and use my instinct instead.  I didn't break the lines because I felt 4 extra characters shouldn't cause an extra line to be created, and I didn't like the alternatives:

I don't like assignments starting with a line continuation:
```python
cache_key = \
    '{}.{}'.format(buf.name,
                   word[:self.complete_min_length].lower())
```

I don't like it when attribute/property access starts a line:
```python
cache_key = '{}.{}' \
            .format(buf.name,
                    word[:self.complete_min_length].lower())
```

and:
```python
cache_key = '{}.{}'.format(buf.name,
                           word[:self.complete_min_length]
                           .lower())
```

The last example is the most acceptable, but it's still only 4 characters over.  If you prefer strict adherence to the default flake8 rules, I'll fix it.